File: baekjoon-online-judge/normal/02533.py
from sys import stdin, setrecursionlimit
setrecursionlimit(int(1e6))
input = stdin.readline

EARLY = 0
GENERAL = 1

def dfs(i):
    visited[i] = True
    dp[EARLY][i] = 1
    for v in graph[i]:
        if not visited[v]:
            dfs(v)
            dp[EARLY][i] += min(dp[EARLY][v], dp[GENERAL][v])
            dp[GENERAL][i] += dp[EARLY][v]

if __name__ == '__main__':
    # Input
    N = int(input())

    graph = [[] for _ in range(N + 1)]
    visited = [False] * (N + 1)
    for _ in range(N - 1):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    # Solve
    dp = [[0] * (N + 1) for _ in range(2)]
    dfs(1)

    # Output
    print(min(dp[EARLY][1], dp[GENERAL][1]))

File: baekjoon-online-judge/gold/02533.java
import java.io.*;
import java.util.*;

public class Main {
    static List<List<Integer>> graph = new ArrayList<>();
    static int[][] dp;
    static boolean[] visited;
    static final int EARLY = 1, GENERAL = 0;
    private static void dfs(int cur) {
        visited[cur] = true;
        dp[EARLY][cur] = 1;
        for (int next : graph.get(cur)) {
            if (!visited[next]) {
                dfs(next);
                dp[EARLY][cur] += Math.min(dp[EARLY][next], dp[GENERAL][next]);
                dp[GENERAL][cur] += dp[EARLY][next];
            }
        }
    }
    public static void main(String[] args) throws IOException {
        // Init
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        // Input
        final int N = Integer.parseInt(br.readLine());
        dp = new int[2][N + 1];
        visited = new boolean[N + 1];
        for (int i = 0; i <= N; ++i) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < N - 1; ++i) {
            st = new StringTokenizer(br.readLine());
            final int u = Integer.parseInt(st.nextToken());
            final int v = Integer.parseInt(st.nextToken());
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        // Solve
        final int root = 1;
        dfs(root);

        // Output
        System.out.println(Math.min(dp[EARLY][root], dp[GENERAL][root]));
    }
}